#### Assignment 1
Add a new column named "AreaPopu" which is equal to the addition of 'Area(Km2)' and 'Population'

In [2]:
import pandas as pd
import numpy as np

In [3]:
pop_df= pd.read_excel('2006.xlsx')

pop_df.head()

STATES  AREA (km2)   Population
0       Abia State         6320     2845380
1    Adamawa State        36917     3178950
2  Akwa Ibom State         7081     3178950
3    Anambra State         4844     4177828
4     Bauchi State        45837     4653066

In [4]:
pop_df['AreaPopu']= pop_df["AREA (km2) "] + pop_df.Population
pop_df.head()

STATES  AREA (km2)   Population  AreaPopu
0       Abia State         6320     2845380   2851700
1    Adamawa State        36917     3178950   3215867
2  Akwa Ibom State         7081     3178950   3186031
3    Anambra State         4844     4177828   4182672
4     Bauchi State        45837     4653066   4698903

### Assignment 3

With pandas groupby method, count all states that belongs to each region

In [38]:
#Creating those a function that can map states to different regions, we have:
def region(state):
    NC = ['Benue State', 'Kogi', 'Kwara', 'Nasarawa', 'Niger', 'Plateau', 'FCT']
    SS = ['Akwa', 'Bayelsa', 'Cross River', 'Rivers', 'Delta', 'Edo']
    SE = ['Abia', 'Anambra', 'Ebonyi', 'Enugu', 'Imo']
    SW = ['Ekiti', 'Lagos', 'Ogun', 'Ondo', 'Osun', 'Oyo']
    NE = ['Adamawa', 'Bauchi', 'Borno', 'Gombe', 'Taraba', 'Yobe']
    NW = ['Jigawa', 'Kaduna', 'Kano', 'Katsina', 'Kebbi', 'Sokoto', 'Zamfara']
    if state in NC:
        return 'North Central'
    elif state in SS:
        return 'South South'
    elif state in SE:
        return 'South East'
    elif state in SW:
        return 'South West'
    elif state in NE:
        return 'North East'
    else:
        return 'North West'

In [9]:
pop_df['STATE'] = [state.split()[0] if len(state.split())<=2 else (state.split()[0] + ' ' + state.split()[1]) for state in (pop_df['STATES'])]

pop_df.head()

STATES  AREA (km2)   Population  AreaPopu      STATE
0       Abia State         6320     2845380   2851700       Abia
1    Adamawa State        36917     3178950   3215867    Adamawa
2  Akwa Ibom State         7081     3178950   3186031  Akwa Ibom
3    Anambra State         4844     4177828   4182672    Anambra
4     Bauchi State        45837     4653066   4698903     Bauchi

In [10]:
pop_df.drop('STATES', axis=1, inplace=True)

In [39]:
#Applying th efunction using pandas.apply() method, we have:

pop_df['Region'] = pop_df['STATE'].apply(region)

#Lets see whats in the new feature and the entire dataset
pop_df.head()

AREA (km2)   Population  AreaPopu      STATE      Region
0         6320     2845380   2851700       Abia  South East
1        36917     3178950   3215867    Adamawa  North East
2         7081     3178950   3186031  Akwa Ibom  North West
3         4844     4177828   4182672    Anambra  South East
4        45837     4653066   4698903     Bauchi  North East

In [40]:
temp = pop_df.groupby(['Region'])
temp.describe(include='all')

AREA (km2)                                               \
                    count unique top freq          mean           std   
Region                                                                  
North Central         6.0    NaN NaN  NaN  34727.666667  22741.089566   
North East            6.0    NaN NaN  NaN  45399.166667  17400.536859   
North West            9.0    NaN NaN  NaN  28578.333333  11816.385551   
South East            5.0    NaN NaN  NaN   5905.000000    876.260806   
South South           5.0    NaN NaN  NaN  15501.200000   4292.906789   
South West            6.0    NaN NaN  NaN  13277.500000   9057.851373   

                                                    ...  STATE               \
                   min       25%      50%      75%  ... unique          top   
Region                                              ...                       
North Central   7315.0  27796.00  30373.0  35347.0  ...      6     Nasarawa   
North East     18768.0  39063.25  45669.5  52314.0  ...      6         Yobe   
North West      7081.0  23154.00  25973.0  36800.0  ...      9       Sokoto   
South East      4844.0   5530.00   5670.0   6320.0  ...      5        Enugu   
South South    10773.0  11077.00  17698.0  17802.0  ...      5  Cross River   
South West      3345.0   7077.50  12375.5  16446.5  ...      6        Lagos   

                                                       
              freq mean  std  min  25%  50%  75%  max  
Region                                                 
North Central    1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
North East       1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
North West       1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
South East       1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
South South      1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  
South West       1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[6 rows x 44 columns]

#### NC = 6, NW=9, NE= 6, SE=5, SS = 5, SW= 6

#### Assignment 3

- From the descriptive Statiatics above, explain at least 3 points
- Generate the Descriptive statistics such that all features (Numerical and Categorical) will be included

In [26]:
data = pd.read_csv('Data.csv')
data.describe()

User ID         Age         Salary
count  4.000000e+02  400.000000     400.000000
mean   1.569154e+07   37.655000   69742.500000
std    7.165832e+04   10.482877   34096.960282
min    1.556669e+07   18.000000   15000.000000
25%    1.562676e+07   29.750000   43000.000000
50%    1.569434e+07   37.000000   70000.000000
75%    1.575036e+07   46.000000   88000.000000
max    1.581524e+07   60.000000  150000.000000

#### From the descriptive Statiatics above, explain at least 3 points

i. All the columns (User ID, Age, Salary) has entries in each row with a total of 400 rows i.e No missing value.

ii. Salary column has a maximum value of 150000 and minimum value of 15000.

iii. Agea has an average of 37.655000

#### Generate the Descriptive statistics such that all features (Numerical and Categorical) will be included

In [29]:
data.describe(include = 'all')

User ID Country  Gender         Age         Salary Purchased
count   4.000000e+02     400     400  400.000000     400.000000       400
unique           NaN      10       2         NaN            NaN         2
top              NaN   Spain  Female         NaN            NaN        No
freq             NaN     266     204         NaN            NaN       257
mean    1.569154e+07     NaN     NaN   37.655000   69742.500000       NaN
std     7.165832e+04     NaN     NaN   10.482877   34096.960282       NaN
min     1.556669e+07     NaN     NaN   18.000000   15000.000000       NaN
25%     1.562676e+07     NaN     NaN   29.750000   43000.000000       NaN
50%     1.569434e+07     NaN     NaN   37.000000   70000.000000       NaN
75%     1.575036e+07     NaN     NaN   46.000000   88000.000000       NaN
max     1.581524e+07     NaN     NaN   60.000000  150000.000000       NaN

### Assignment 4

Using the same logic that was used above in

> print(Not_buy,'(',(Not_buy/len(data)) * 100, '%) of the Customers did not buy the Ads')

What is the number and percentage of customers who did not buy the ADs?

In [41]:
# Mean Satistics of Age and Salary of Customers that did not buy and those that bought the Ads
Purchased_or_not = data.groupby('Purchased')
Purchased_or_not.mean()

User ID        Age        Salary
Purchased                                       
No         1.569116e+07  32.793774  60544.747082
Yes        1.569222e+07  46.391608  86272.727273

In [46]:
# ratio of users that did bought the Ads were only 36%
data.Purchased.value_counts()
Not_buy = data.Purchased.value_counts()[0]
buy  = data.Purchased.value_counts()[1]
print(Not_buy,'(',(Not_buy/len(data)) * 100, '%) of the Customers did not buy the Ads')

257 ( 64.25 %) of the Customers did not buy the Ads


### Assignment 7

Rename the **'Yes'** column as **'Purchased'**

In [47]:
#If the dataset is in the same directory with the notebook 

dt = pd.read_csv('Data_Miss.csv')
dt

Country   Age   Salary Purchased
0   France  44.0  72000.0        No
1    Spain  27.0  48000.0       Yes
2  Germany  30.0  54000.0        No
3    Spain  38.0  61000.0        No
4  Germany  40.0      NaN       Yes
5   France  35.0  58000.0       Yes
6    Spain   NaN  52000.0        No
7   France  48.0  79000.0       Yes
8  Germany  50.0  83000.0        No
9   France  37.0  67000.0       Yes

In [49]:
dt.isnull().sum()

Country      0
Age          1
Salary       1
Purchased    0
dtype: int64

In [50]:
#Fill the missing value with mean Age
dt.Age.fillna(dt.Age.mean())

0    44.000000
1    27.000000
2    30.000000
3    38.000000
4    40.000000
5    35.000000
6    38.777778
7    48.000000
8    50.000000
9    37.000000
Name: Age, dtype: float64

In [51]:
#Encoding features one by one, then concatenate 
enc_purchased = pd.get_dummies(dt['Purchased'] , drop_first = True)
Enc_Country = pd.get_dummies(dt['Country']) 

#Lets sees what we have in the encoded contry data
Enc_Country

France  Germany  Spain
0       1        0      0
1       0        0      1
2       0        1      0
3       0        0      1
4       0        1      0
5       1        0      0
6       0        0      1
7       1        0      0
8       0        1      0
9       1        0      0

In [52]:
final_df = pd.concat([dt, enc_purchased, Enc_Country],axis=1)
final_df

Country   Age   Salary Purchased  Yes  France  Germany  Spain
0   France  44.0  72000.0        No    0       1        0      0
1    Spain  27.0  48000.0       Yes    1       0        0      1
2  Germany  30.0  54000.0        No    0       0        1      0
3    Spain  38.0  61000.0        No    0       0        0      1
4  Germany  40.0      NaN       Yes    1       0        1      0
5   France  35.0  58000.0       Yes    1       1        0      0
6    Spain   NaN  52000.0        No    0       0        0      1
7   France  48.0  79000.0       Yes    1       1        0      0
8  Germany  50.0  83000.0        No    0       0        1      0
9   France  37.0  67000.0       Yes    1       1        0      0

In [55]:
final_df.rename(columns={'Yes':'Purchased'})
final_df

Country   Age   Salary Purchased  Yes  France  Germany  Spain
0   France  44.0  72000.0        No    0       1        0      0
1    Spain  27.0  48000.0       Yes    1       0        0      1
2  Germany  30.0  54000.0        No    0       0        1      0
3    Spain  38.0  61000.0        No    0       0        0      1
4  Germany  40.0      NaN       Yes    1       0        1      0
5   France  35.0  58000.0       Yes    1       1        0      0
6    Spain   NaN  52000.0        No    0       0        0      1
7   France  48.0  79000.0       Yes    1       1        0      0
8  Germany  50.0  83000.0        No    0       0        1      0
9   France  37.0  67000.0       Yes    1       1        0      0

### Assignment 8

Examine the pairwise relationsip between other variables and give your findings eg

- Age and Salary
- Country_Spain and Salary

In [56]:
# The correlation coefficient of numeric the features
corre = final_df.corr()
#We can view the correlation table
corre

Age    Salary       Yes    France   Germany     Spain
Age      1.000000  0.982495 -0.212359  0.274011  0.119144 -0.462602
Salary   0.982495  1.000000 -0.060157  0.403914  0.218273 -0.618261
Yes     -0.212359 -0.060157  1.000000  0.408248 -0.218218 -0.218218
France   0.274011  0.403914  0.408248  1.000000 -0.534522 -0.534522
Germany  0.119144  0.218273 -0.218218 -0.534522  1.000000 -0.428571
Spain   -0.462602 -0.618261 -0.218218 -0.534522 -0.428571  1.000000

i. The relationship beteween Age and salary has a very strong positive correlation since the value is close to +1. This implies that if age increases, salary also increases.

ii.The relationship beteween Spain and salary has a very strong negative correlation since the value is close to -1. This implies that if value in Spain increases, salary also decreases and vise versa. 